### PR TITLE
[#1221] feat(rust): Add grpc graceful shutdown

### PR DESCRIPTION
**Pull request by wenlongbrother**
### What changes were proposed in this pull request?

Add grpc gracful shutdown

### Why are the changes needed?

**Data Integrity**: To ensure that any ongoing write operations or transactions are properly handled to prevent data corruption or loss.

**Resource Cleanup**: To release resources such as file handles, network connections, and memory to prevent resource leaks.

**User Experience**: To provide feedback to users, where possible, informing them that the program is shutting down, to avoid sudden service interruptions causing distress.

**State Preservation**: In some applications, it may be necessary to save the current state so that the next start can continue from the same point.

**Dependency Service Coordination**: If your service is a dependency for other services, a graceful shutdown can ensure dependent services are notified to avoid causing cascading failures.

**Error Tracking and Logging**: Recording sufficient information during the shutdown process is very helpful for subsequent error analysis and troubleshooting.

**Compliance with Laws and Regulations**: Some applications process sensitive data and need to ensure compliance with legal and regulatory requirements to prevent data breaches during the shutdown process.

In distributed systems and microservices architectures, graceful shutdown becomes particularly important, as the interactions between services are more complex, and improper shutdowns can trigger a range of issues. For example, if a database service shuts down abruptly without gracefully handling current operations, it could lead to data inconsistencies that affect the stability of the entire system.

Overall, graceful shutdown is about ensuring the robustness and reliability of systems, maintaining service quality, and protecting data security even when service termination is necessary.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

image

image

### CODE REVIEW

File: rust/experimental/server/src/lib.rs [hunk 1]
```diff
@@ -75,8 +81,21 @@ pub async fn start_uniffle_worker(config: config::Config) -> Result<()> {
...
             .max_decoding_message_size(usize::MAX)
             .max_encoding_message_size(usize::MAX);
-        let _ = Server::builder().add_service(service).serve(addr).await;
+        let _ = Server::builder().add_service(service).serve_with_shutdown(addr, async {
+            rx.await.expect("graceful_shutdown fail");
+            println!("Successfully received the shutdown signal.");
+        }).await;
     });
+
+    tokio::spawn(async move {
```
Comment (zuston) [inline rust/experimental/server/src/lib.rs:90]: The concrete runtime has been wrapped in the runtime manager. Please respect this.

File: rust/experimental/server/src/lib.rs [hunk 2]
```diff
@@ -17,6 +17,30 @@
 
 #![feature(impl_trait_in_assoc_type)]
 
+use std::net::{IpAddr, Ipv4Addr, SocketAddr};
```
Comment (zuston) [inline rust/experimental/server/src/lib.rs:20]: I'm not sure whether we should have the rule to decide the module importing sequence rule. WDYT?

File: rust/experimental/server/src/main.rs [hunk 3]
```diff
@@ -231,11 +235,17 @@ fn main() -> Result<()> {
                 AWAIT_TREE_REGISTRY.clone(),
             )))
             .add_service(service)
-            .serve(addr)
+            .serve_with_shutdown(addr, async {
+                if let Err(err) = rx.await {
+                    error!("Worker stopping error: {:?}.", err);
```
Comment (zuston) [inline rust/experimental/server/src/main.rs:240]: How about `Errors on stopping the GRPC service, err: {:?}`

File: rust/experimental/server/src/main.rs [hunk 4]
```diff
@@ -231,11 +235,17 @@ fn main() -> Result<()> {
...
+                } else {
+                    info!("Worker stopped.");
```
Comment (zuston) [inline rust/experimental/server/src/main.rs:242]: How about `GRPC service has been graceful stopped.`

File: rust/experimental/server/src/signal.rs [hunk 5]
```diff
@@ -19,6 +19,7 @@
 pub mod details {
     use log::info;
     use signal_hook::{consts::TERM_SIGNALS, iterator::Signals};
+    use tokio::sync::oneshot::Sender;
 
     pub fn wait_for_signal() {
```
Comment (zuston) [inline rust/experimental/server/src/signal.rs:24]: This could be deleted.

**Review by zuston — COMMENTED** [review]
Great work! @wenlongbrother 

Overall LGTM. left minor comments.

**Comment by zuston** [comment]
BTW, please run the `cargo fmt` to format your code

File: rust/experimental/server/src/lib.rs (hunk 2 above)
Comment (wenlongbrother) [reply rust/experimental/server/src/lib.rs:20]: sorry, just using RustRover to format code manually

File: rust/experimental/server/src/lib.rs (hunk 1 above)
Comment (wenlongbrother) [reply rust/experimental/server/src/lib.rs:90]: I see default_runtime has default_runtime, it's for not important tasks, how about use it?

**Comment by smallzhongfeng** [comment]
> BTW, please run the `cargo fmt` to format your code

Maybe we can add this  `cargo fmt`  to ci ?

**Comment by wenlongbrother** [comment]
Hi @zuston , I push some new changes, can you take a look again, thank you for your review.

**Comment by zuston** [comment]
> > BTW, please run the `cargo fmt` to format your code
> 
> Maybe we can add this `cargo fmt` to ci ?

 fmt check has been added into CI. but the formatting code for developer is not suitable.

**Review by zuston — APPROVED** [review]

**Comment by zuston** [comment]
Thanks for your contribution! @wenlongbrother

**Pull request merged**